bokehjs/src/coffee/models/tools/draw/line_edit_tool.ts

```diff
@@ -0,0 +1,78 @@
+import * as p from "core/properties"
+import {ColumnDataSource} from "models/sources/column_data_source"
+import {SelectTool} from "../gestures/select_tool"
+import {EditToolView} from "./edit_tool"
+
+export interface BkEv {
+  bokeh: {
+    sx: number
+    sy: number
+  }
+  srcEvent: {
+    shiftKey?: boolean
+  }
+  keyCode: number
+  timeStamp: number
+}
+
+export class LineEditToolView extends EditToolView {
+  model: LineEditTool
+  _basepoint: [number, number]
+
+  _pan_start(e: BkEv): void {
+    // Perform hit testing
+    this._select_event(e, false);
+    const point = this._map_drag(e.bokeh.sx, e.bokeh.sy);
+    if (point == null) {
+      return
+    }
+    this._basepoint = point;
+  }
+
+  _pan(e: BkEv): void {
+    const ds = this.model.source;
+    const point = this._map_drag(e.bokeh.sx, e.bokeh.sy)
+    if (!ds.selected['1d'].indices.length || point == null || this._basepoint == null) {
+      return;
+    }
+    const [x, y] = point
+	const [px, py] = this._basepoint
+    const [dx, dy] = [x-px, y-py];
+    this._basepoint = point;
+    const index = ds.selected['1d'].indices[0];
+    const xs = ds.data[this.model.x][index];
+    const ys = ds.data[this.model.y][index];
+    for (let i = 0; i < ys.length; i++) {
+      xs[i] = xs[i]+dx;
+      ys[i] = ys[i]+dy;
+    }
+    ds.change.emit(undefined);
+  }
+
+  _pan_end(_e: BkEv): void {
+    this.model.source.selected['1d'].indices = [];
+    this._basepoint = null;
```

Comment:
Given this assignment earlier definition of `_basepoint` should be changed to `_basepoint: [number, number] | null`.

Reply:
Thanks, btw, should all private variables be declared/typed in this way? I tried to look at other classes for examples, but the conversion is probably not complete yet so not all classes are doing it.

Reply:
This is unrelated to whether a variable is private or not, but to make sure that the type reflects all allowed values. If this line wouldn't exist, then it would be sufficient to use the existing type. We use almost the strictest possible settings (almost because there are a few new options which we didn't exercise yet), which don't allow implicit `null` or `undefined` values. Every use of these must be reflected in the type of a variable, signature, etc. One gray and unexplored area so far is analysis whether an instance variable, like `_basepoint`, were initialized or not. This can result in a behavior that would indicate its type to be `something | undefined` (`[number, number] | undefined` in this case), which is not. I will enable feature checking initialization status at some point, but for now it would only add another 1000 or so errors to the existing long list.

Reply:
I'm actually getting ahead of myself, because was speaking about features not released yet. This will be available in TypeScript 2.7 (sometime this month).